src/pybind/mgr/smb/resources.py

```diff
@@ -191,15 +191,43 @@ def with_qos_updates(
             'read_bw_limit': read_bw_limit,
             'write_bw_limit': write_bw_limit,
         }
-        updates = {k: v for k, v in updates.items() if v is not None}
-        new_values = {**current, **updates}
+
+        new_values = {
+            **current,
+            **{k: v for k, v in updates.items() if v is not None},
+        }
+
+        # Check if all QoS values are being explicitly set to 0 (disabled)
+        all_zero = all(
```

Comment:
I find this even more difficult to read and understand than the initial implementation of this function. If my suggestion there is insufficient - let me know and let's talk.